### PR TITLE
Declare SiTrackerMultiRecHit::cloneSH() to override

### PR DESCRIPTION
**Pull request by makortel**
#### PR description:

Fixes clang warnings in `Validation/SiPixelPhase1HitsV`, `Validation/SiPixelPhase1RecHitsV`, and `Validation/SiPixelPhase1TrackingParticleV`.

#### PR validation:

Code compiles.

### CODE REVIEW

**Comment by makortel** [comment]
@cmsbuild, please test

File: DataFormats/TrackerRecHit2D/interface/SiTrackerMultiRecHit.h [hunk 1]
```diff
@@ -24,7 +24,7 @@ class SiTrackerMultiRecHit : public BaseTrackerRecHit {
 
   SiTrackerMultiRecHit* clone() const override { return new SiTrackerMultiRecHit(*this); }
 #ifdef NO_DICT
```
Comment (makortel) [inline DataFormats/TrackerRecHit2D/interface/SiTrackerMultiRecHit.h]: The `NO_DICT` appears to be defined only here
https://github.com/cms-sw/cmssw/blob/fb6a9561605932038903010b6aff187d6883e6e2/CondFormats/SiPixelObjects/interface/SiPixelFedCablingMap.h#L14-L16
and now that header gets somehow included e.g. in `Validation/SiPixelPhase1HitsV/src/SiPixelPhase1HitsV.cc` (with a little effort I wasn't able to figure out how).

I believe those checks are no longer need (at least here, not fully sure if in `CondFormats`). If people agree, I could remove the `NO_DICT` here in this PR.

Comment (slava77) [reply DataFormats/TrackerRecHit2D/interface/SiTrackerMultiRecHit.h]: looking at other cases of cloneSH, I'm guessing that here we need ```#ifndef __GCCXML__```.
Apparently this one was missed in https://github.com/cms-sw/cmssw/commit/8e55c0c6f22f00d47ec09b46f7422a4d00a36ee7

Comment (makortel) [reply DataFormats/TrackerRecHit2D/interface/SiTrackerMultiRecHit.h]: I suspect none of the `#ifndef __CINT__`, `__MAKECINT__`, `__REFLEX__`, `__GCCXML__` are needed anymore with ROOT 6.

Comment (slava77) [reply DataFormats/TrackerRecHit2D/interface/SiTrackerMultiRecHit.h]: perhaps then it's worth to cleanup __GCCXML__ in more places, at least for the tracking rechits

Comment (makortel) [reply DataFormats/TrackerRecHit2D/interface/SiTrackerMultiRecHit.h]: I agree (and just created an issue #28748). Would you prefer that such a cleanup would be done in a separate PR for e.g. all tracking rechits (or reco dataformats) over removing the `#ifdef NO_DICT` here.

Comment (slava77) [reply DataFormats/TrackerRecHit2D/interface/SiTrackerMultiRecHit.h]: IIUC, this PR is technically complete and it looks like the idea is to make a larger cleanup separately.
So, we could go ahead with this PR as is.

Comment (makortel) [reply DataFormats/TrackerRecHit2D/interface/SiTrackerMultiRecHit.h]: Fine for me.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/28743	5b59a35 
- technical; in line with the PR description and discussion
- jenkins tests pass.

@makortel in case you decide to update here, it'd be fine with me

**Comment by makortel** [comment]
> in case you decide to update here, it'd be fine with me

Just to make it clear, I won't :)

**Pull request merged**